baselines/tamuna/tamuna/main.py

```diff
@@ -0,0 +1,203 @@
+from tamuna.dataset import load_datasets
+from tamuna.strategy import CentralizedFedAvgStrategy, TamunaStrategy
+from tamuna.utils import compare_histories, save_results_as_pickle
+
+
+@hydra.main(config_path="conf", config_name="base", version_base=None)
+def main(cfg: DictConfig) -> None:
+    """Run CNN federated learning on MNIST.
+
+    Parameters
+    ----------
+    cfg : DictConfig
+        An omegaconf object that stores the hydra config.
+    """
+    # print config structured as YAML
+    print(OmegaConf.to_yaml(cfg))
+
+    # Hydra automatically creates an output directory
+    # Let's retrieve it and save some results there
+    save_path = HydraConfig.get().runtime.output_dir
+    with open(f"{save_path}/config.yaml", "wt") as handle:
+        OmegaConf.save(cfg, handle)
+
+    # partition dataset and get dataloaders
+    trainloaders, testloader = load_datasets(
+        num_clients=cfg.server.num_clients, iid=cfg.dataset.iid
+    )
+
+    tamuna_histories = run_tamuna(cfg, save_path, testloader, trainloaders)
+    fedavg_histories = run_fedavg(cfg, save_path, testloader, trainloaders)
+
+    with open("model_dim.txt", "rt") as handle:
+        dim = int(handle.readline())
+
+    compare_histories(tamuna_histories, fedavg_histories, dim, save_path, cfg)
+
+
+def run_fedavg(
+    cfg: DictConfig,
+    save_path: str,
+    testloader: DataLoader,
+    trainloaders: List[DataLoader],
+) -> List[History]:
+    """Run FedAvg.
+
+    Parameters
+    ----------
+    cfg : DictConfig
+        An omegaconf object that stores the hydra config.
+    save_path: str
+         Path where to save the results of the training.
+    testloader: DataLoader
+        Dataloader contaning test split for centralized evaluation
+    trainloaders: List[DataLoader]
+        Dataloaders for clients
+
+    Returns
+    -------
+    List[History]
+        Histories of every run (based of n_repeats).
+    """
+    # prepare function that will be used to spawn each client
+    client_fn = client.gen_fedavg_client_fn(
+        trainloaders=trainloaders,
+        learning_rate=cfg.client.learning_rate,
+        model=cfg.model,
+        client_device=cfg.client.client_device,
+    )
+
+    # using only central evaluation
+    evaluate_fn = server.gen_evaluate_fn(
+        testloader, device=cfg.server.server_device, model=cfg.model
+    )
+
+    histories = []
+    for i in range(cfg.meta.n_repeats):
+        np.random.seed(cfg.meta.seed)
+        torch.manual_seed(cfg.meta.seed)
+
+        # Start simulation
+        history = fl.simulation.start_simulation(
+            client_fn=client_fn,
+            num_clients=cfg.server.num_clients,
+            config=fl.server.ServerConfig(num_rounds=cfg.server.num_rounds),
+            client_resources={
+                "num_cpus": cfg.client.client_resources.num_cpus,
+                "num_gpus": cfg.client.client_resources.num_gpus,
+            },
+            strategy=CentralizedFedAvgStrategy(
```

Comment:
I wonder if it would be better not to define your own strategy in this case and instead directly use the built-in `FedAvg`. You can of course still override the `evaluate_fn` and the config sent to clients during `configure_fit` list shown below. I think it's preferable to work like this.

```python

# somewhere in your main you can define this
def get_fit_config(is_tamuna: bool):
    def fit_config(server_round: int) -> Dict[str, Scalar]:
        """Return a configuration with static batch size and (local) epochs."""
        if is_tamuna:
            # prepare config
        else:
            local_epochs: Scalar = cfg.server.clients_per_round
            config = {
                "epochs": local_epochs,  # Number of local epochs done by clients
            }
        return config
    return fit_config

strategy = fl.server.strategy.FedAvg(
    ... # add all other relevant arguments,
    evaluate_fn=evaluate_fn,
    on_fit_config_fn=get_fit_config(False),
    )
```

Reply:
I left some other comments in the `strategy.py`

Reply:
Ok, I removed my custom class for FedAvg and wrote it using built-in FedAvg.